pool.go

```diff
@@ -259,49 +254,53 @@ func (p *connPool) Put(cn *conn) error {
 }
 
 func (p *connPool) Remove(cn *conn) error {
-	if p.isClosed() {
-		atomic.AddInt32(&p.size, -1)
-		return cn.Close()
+	if p.closed() {
+		// Close already closed all connections.
+		return nil
 	}
 
-	// Replace existing connection with new one and unblock `wait`.
+	// Replace existing connection with new one and unblock waiter.
 	newcn, err := p.new()
 	if err != nil {
-		atomic.AddInt32(&p.size, -1)
-	} else {
-		p.Put(newcn)
+		return p.conns.Remove(cn)
 	}
-
-	return cn.Close()
+	p.freeConns <- newcn
+	return p.conns.Replace(cn, newcn)
 }
 
-// Len returns number of idle connections.
+// Len returns total number of connections.
 func (p *connPool) Len() int {
-	return len(p.freeConns)
+	return p.conns.Len()
 }
 
-// Size returns number of connections in the pool.
-func (p *connPool) Size() int {
-	return int(atomic.LoadInt32(&p.size))
+// FreeLen returns number of free connections.
+func (p *connPool) FreeLen() int {
+	return len(p.freeConns)
 }
 
-func (p *connPool) Close() (retErr error) {
-	if !atomic.CompareAndSwapInt32(&p.closed, 0, 1) {
-		return nil
+func (p *connPool) Close() error {
+	if !atomic.CompareAndSwapInt32(&p._closed, 0, 1) {
+		return errClosed
```

Comment:
I think I returned nil for a reason here, but I can't remember why. Should be OK though